### DISCUSSION TITLE
Invalid create() Unique constraint failed on the fields: (`id`)

### DISCUSSION THREAD

**Discussion by vladandrei0** [prisma/prisma#5256] [Q&A] [answered]
Hello
I have this model:
```prisma
model Somename {
  id Int @id @default(autoincrement())
  name String
}
```

I get this error when trying to create an entry for the model: Invalid `prisma.somename.create()` Unique constraint failed on the fields: (`id`)

I have previously created 7-8 entries manually in the DB (Postgres). The indexes were like (1,3, 4,5, 9, 11). Is this interfering with autoincrement()?
If I delete the entries or change the existing id values to a higher range it works. So I know the cause of the error now but I want to find out why does it happen? When creating a new entry how does autoincrement works? I could not find any relevant documentation. 

PS: my DB knowledge is at best very limited 🤷‍♀️

All the best!

**Comment by ryands17** [comment, marked as answer]
Hey @vladandrei0 👋 

Autoincrement creates its own sequence starting from 1 which you can read more about [here](https://www.postgresqltutorial.com/postgresql-serial/). 

If you add random records with the id, then Postgres doesn't know that and it tries to start from 1. If 1 is already present, then it will throw an error. 

So the best practice in this case is always let the database generate the `id` for you instead of you adding it manually. Hope this helps :)

**Reply by vladandrei0** [reply]
Yes, thanks!! No more doubts now :))

**Reply by PwntasticKev** [reply]
Thank you kind sir.

**Reply by god-of-js** [reply]
It worked like Magic.

Normally, I know this but I was trying to port existing data with it's own ID.

Anyway, I resorted to placing that ID in a ```uid``` field.

And it worked perfectly

**Reply by Reactplus-Ngoc** [reply]
Yes. Thanks for pointing out. <3

**Reply by Bennison-Devadoss** [reply]
@ryands17 , Before I started to use Prisma, I used sequelize, At that time I have not seen this as an issue in sequelize. If the issue comes from the database this error will be reflected in both sequelize, and prisma.

**Reply by BennisonDevadoss** [reply]
Hey, @vladandrei0, if you utilized the prisma.<model>.upstream() method to generate seed data, it might be the cause of this error. The issue lies in the fact that when running the upstream method, Prisma doesn't generate new values using auto-increment for the primary key id, as the object you intend to store already has the value for the primary key.

**Reply by BennisonDevadoss** [reply]
I encountered a unique constraint error when executing the seed data. If anyone experiences a similar error (specifically, a Unique constraint failure on the 'id' field) within the prisma.upstream() method during a seed generation, we can address it by using the code provided below.
```js
import { cloneDeep } from 'lodash';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

async function main() {
  const Role = {
    User: 'User',
    Admin: 'Admin'
  };

  const userData = [
    {
      id: 1,
      first_name: 'Bennison',
      last_name: 'Devadoss',
      email: 'bennison@your-domain.com',
      role: Role.Admin,
      organization_id: 1,
      created_by: 1,
      created_at: new Date(),
      updated_at: new Date()
    },
    {
      id: 2,
      first_name: 'Sanjay',
      email: 'sanjay@your-domain.com',
      role: Role.Admin,
      organization_id: 1,
      created_by: 1,
      created_at: new Date(),
      updated_at: new Date()
    }
  ];
  for (const user of userData) {
    const createAtts = cloneDeep(user);
    delete createAtts.id;
    await prisma.user.upsert({
      where: {
        id: user.id
      },
      create: { ...createAtts, encrypted_password: 'xxxxxxxxxxxxxxxxxxxxxx' },
      update: { ...user, encrypted_password: 'xxxxxxxxxxxxxxxxxxxxxx' }
    });
  }
}

main()
```

**Reply by jbsmith86** [reply]
Horrible design, inexcusable sloppiness. You create a single record manually and now no person can add records

**Reply by molavec** [reply]
From @Bennison-Devadoss solution, I made a little change that works for me and avoid use of lodash 



```javascript
const transactionResult = await prisma.$transaction(
    userData.map(user=> {
      // const createAtts = cloneDeep(service);
      const { id, ...rest } = user;
      return prisma.user.upsert({
        where: { id: userId },
        update: user,
        create: rest,
      })
    })
  );
```

**Reply by ltann** [reply]
@jbsmith86 why mess with the auto increment column that makes all your records unique ?

**Comment by ajmnz** [comment]
If you have imported records and manually set the `id` on each row, it's possible that Postgres's sequence isn't in sync with your current `id` count. To fix it and be able to insert records again without getting a unique constraint violation, you'll have to reset the internal sequence Postgres uses to keep track of the autoincremental field.

```sql
SELECT setval(pg_get_serial_sequence('tableName', 'id'), coalesce(max(id)+1, 1), false) FROM tableName;
```

More info on [this SO answer](https://stackoverflow.com/a/37972960/10831896).

---

### Note
If your table names start with an uppercase letter, you'll have to add single and double quotes to them inside the `pg_get_serial_sequence` function.

```sql
SELECT setval(pg_get_serial_sequence('"Orders"', 'id'), coalesce(max(id)+1, 1), false) FROM "Orders";
```

**Reply by causztic** [reply]
thank you for this!

**Reply by Sax-Yusuph** [reply]
Thanks. this saved me a ton

**Reply by scotm** [reply]
Indeed, this was a very helpful tip. Thank you so much.

Had to use the $queryRawUnsafe method, since single-quotes broke my usage of $queryRaw.

**Reply by daylonball** [reply]
Thank you so much! Saved me lots of time

**Reply by FlorianWendelborn** [reply]
I ran into this issue when doing a primitive database "dump" by just `tar.gz`ing a remote database to use for local testing. Worked flawlessly!

**Reply by odi89** [reply]
Superb, thank you for this👍

**Reply by jeremyikwuje** [reply]
You just saved my project. I was already thinking of doing a total database reset and buildup as the table is the base table. 

Thank you so much, man.

**Reply by s-1-n-t-h** [reply]
thanks !

**Reply by samolubode** [reply]
Thanks for this solution. I had initially seeded my tables, which led to this error when I was trying to create an additional row.

**Reply by mon-jai** [reply]
Thank you. You saved my day (and my grade too).

**Reply by taylor-lindores-reeves** [reply]
Great advice! I had to import new vehicle makes/models and was getting the unique ID constraint on the makes table. Ran your double quote script against both makes and models tables and it worked like charm, now seeding successfully.

**Reply by KirschX** [reply]
I love you, you really save my proejct!

**Reply by TheRealSmeddy** [reply]
Hey, thanks a lot!

**Reply by guisantosalves** [reply]
nice, it helps me a lot

**Reply by liunian321** [reply]
非常感谢

**Reply by hilmarch-dev** [reply]
nice

**Comment by lostation** [comment]
Hi dudes...

I came across the same bad experience...but I have an issue with the solution...

Can you give me advice ? and a working example ?

Currently I'm doing this line of code:

this.prisma.$executeRawUnsafe(`SELECT setval(pg_get_serial_sequence('"Tag"', "id"), coalesce(max(id)+1, 1), false) FROM "Tag";`);


but it returns me error like:
```
Invalid `prisma.$executeRaw()` invocation: ...
{
  code: 'P2010',
  clientVersion: '4.8.1',
  meta: {
    code: '42883',
    message: 'db error: ERROR: function pg_get_serial_sequence(unknown, bigint) does not exist\n' +
      'HINT: No function matches the given name and argument types. You might need to add explicit type casts.'
  },
  batchRequestIdx: undefined
}
```

What should I change to fix that ? Thanks in advance.
Lo.

**Reply by lostation** [reply]
Ok with that right syntax from typescript code:

```
this.prisma.$executeRaw`SELECT setval(pg_get_serial_sequence('"Tag"', 'id'), coalesce(max(id)+1, 1), false) FROM "Tag";`;
```

**Reply by lostation** [reply]
Or:
```
        console.log("Sync Tag id sequence",
            await this.prisma.$executeRaw`
            SELECT setval((
                SELECT PG_GET_SERIAL_SEQUENCE('"Tag"', 'id')),
                (SELECT (MAX("id") + 1) FROM "Tag"),
                false) FROM "Tag";
            `);
```

**Reply by paiameya** [reply]
I literally ran this command in psql 
`
SELECT setval(pg_get_serial_sequence('"Activity"', 'id'), coalesce(max(id)+1, 1), false) FROM "Activity";
`

**Comment by xulilu** [comment]
I was able to reset the id_seq by  `ALTER SEQUENCE "uploads_id_seq" RESTART WITH 100;`
Learned from https://stackoverflow.com/a/78208131
